src/main/scala/com/microsoft/ml/spark/vw/VowpalWabbitBase.scala

```diff
@@ -0,0 +1,421 @@
+    val no_stdin = "--no_stdin"
+    if (args.indexOf(no_stdin) == -1)
+      args.append(" ").append(no_stdin)
+
+    // need to keep reference around to prevent GC and subsequent file delete
+    if (getEnableCacheFile) {
+      val cacheFile = java.io.File.createTempFile("vowpalwabbit", ".cache")
+      cacheFile.deleteOnExit()
+
+      args.append(s" -k --cache_file=${cacheFile.getAbsolutePath} --passes ${getNumPasses}")
+    }
+
+    log.info(s"VowpalWabbit args: ${args}")
+
+    args.result
+  }
+
+  /**
+    * Internal training loop.
+    * @param df the input data frame.
+    * @param vwArgs vw command line arguments.
+    * @param contextArgs This lambda returns command line arguments that are executed in the final execution context.
+    *                    It is used to get the partition id.
+    */
+  private def trainInternal(df: DataFrame, vwArgs: String, contextArgs: => String = "") = {
+    val applyLabel = createLabelSetter(df)
+
+    val featureColIndices = generateNamespaceInfos(df.schema)
+
+    // only perform the inner-loop if we cache the inputRows
+    val numPassesInSpark = if (!getEnableCacheFile && getCacheRows) getNumPasses else 1
+
+    def trainIteration(inputRows: Iterator[Row],
+                       localInitialModel: Option[Array[Byte]],
+                       pass: Int): Iterator[Option[Array[Byte]]] = {
+
+      // potential cache data
+      val iteratorGenerator = createDataIterator(inputRows, numPassesInSpark)
+
+      // construct command line arguments
+      val args = buildCommandLineArguments(vwArgs, contextArgs)
+
+      StreamUtilities.using(if (localInitialModel.isEmpty) new VowpalWabbitNative(args)
+        else new VowpalWabbitNative(args, localInitialModel.get)) { vw =>
+        StreamUtilities.using(vw.createExample()) { ex =>
+
+            // loop in here to avoid
+            // - passing model back and forth
+            // - re-establishing network connection
+            for (_ <- 0 until numPassesInSpark) {
+              // pass data to VW native part
+              for (row <- iteratorGenerator()) {
+                // transfer label
+                applyLabel(row, ex)
+
+                // transfer features
+                for (ns <- featureColIndices)
+                  row.get(ns.colIdx) match {
+                    case dense: DenseVector => ex.addToNamespaceDense(ns.featureGroup,
+                      ns.hash, dense.values)
+                    case sparse: SparseVector => ex.addToNamespaceSparse(ns.featureGroup,
+                      sparse.indices, sparse.values)
+                  }
+
+                ex.learn
+                ex.clear
+              }
+
+              vw.endPass
+            }
+
+            if (getEnableCacheFile)
+              vw.performRemainingPasses
+          }
+
+          // only export the model on the first partition
+          Seq(if (TaskContext.get.partitionId == 0) Some(vw.getModel) else None).iterator
```

Comment:
will this work with empty partitions - and also if local initial model is empty:
if (localInitialModel.isEmpty) new VowpalWabbitNative(args)